### ISSUE TITLE
Cross-cluster Raft network is deployed using Kafka consensus?

### ISSUE THREAD

**Issue by mr-rooftop** [APGGroeiFabriek/PIVT#38]
First of all, thanks for this great repository! It helped me in the past setting up HF networks on K8s, and the new addition of the cross-cluster network looks very juicy :pray: 

I'm currently trying to deploy the network exactly as described in the Cross-cluster Raft network section. 

Now that I launched the network (https://github.com/APGGroeiFabriek/PIVT#launch-the-network), I look at the deployed pods and see that there are Kafka and Zookeeper instances. 
Furthermore I looked at what's actually created on the cluster via helm template
`helm template ./hlf-kube --name hlf-kube -f samples/cross-cluster-raft-tls/cluster-one/network.yaml -f samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml --set peer.launchPods=false --set orderer.launchPods=false`
and it says that KAFKA is used instead of RAFT consensus.

Is this on purpose? Do I overlook something? I copied all commands from the README directly to my console. 
I'm using Helm 2.11.0, Kubectl 1.15.5 and the K8s cluster is v1.15.9.

Best Regards

**Comment by LudoNieuw** [comment]
Hi there,

Can you check your network.yaml file for cluster one? It should have ‘OrdererType: etcdraft’ for raft ordered to be deployed.

Regards,

Ludo

**Comment by mr-rooftop** [comment]
Thanks for the fast response!

As far as I know, you only specify `OrdererType: etcdraft` in configtx.yaml, not in network.yaml, isn't it?

In configtx.yaml `OrdererType: etcdraft` is specified yes.

**Comment by LudoNieuw** [comment]
Yeah, not network but configtx.yaml indeed :) 

Did you run ‘helm del —purge hlf-kube’ before setting up the cross-cluster thing? Perhaps you still have some Kafka resources running from an earlier deployment?

I just tried the command you described with helm 2.11 and kube-client 1.15.0 and live-server 1.14.8 and did not see any Kafka resources.


Verstuurd vanaf smart phone

**Comment by mr-rooftop** [comment]
I just created a new cluster, and tried it again by just copiyng the commands as suggested. Again Kafka instances, no idea why, really strange.

Thanks for trying it out, good to know that it works on your side.

What are actually the helm, kubectl and k8s server versions that you would suggest for smooth sailing?

**Comment by raftAtGit** [comment]
`Kafka` and `Zookeeper` pods should be created only when `hlf-kafka.enabled=true`, can you double check that value in `values.yaml` file?

`OrdererType: kafka` in `configtx.yaml` does not determine if Kafka pods are created, but of course should be in synch with `hlf-kafka.enabled`

**Comment by mr-rooftop** [comment]
Thanks for pointing that out @raftAtGit , I didn't know that only the values.yaml is responsible for Kafka and Zookeeper Pods.

Your repsonses are wonderful, thank you so much :pray: 
It's rare to see such active support in HF projects!


I assume you are referring to the values.yaml file as in fabric-kube/hlf-kube/values.yaml. 
Yes hlf-kafka.enabled was always set to false as in 
`hlf-kafka:
  enabled: false`


I cloned the PIVT repo again, and created two new k8s clusters. On one I tried to follow your cross-cluster section with helm 2.11.0, and on the other with helm 2.15.2. 


I never did `helm dependency update ./hlf-kube/` as you suggest in Prerequisites of Launching the network (https://github.com/APGGroeiFabriek/PIVT#launching-the-network), because I realized that the only requirements are Kafka (hlf-kube/requirements.yaml):

`dependencies:
  - name: kafka
    alias: hlf-kafka
    repository: http://storage.googleapis.com/kubernetes-charts-incubator
    version: 0.20.8
    condition: hlf-kafka.enabled
`

When I tried to install the helm chart for cluster one with `helm install ./hlf-kube --name hlf-kube -f samples/cross-cluster-raft-tls/cluster-one/network.yaml -f samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml --set peer.launchPods=false --set orderer.launchPods=false`
I got 
`Error: found in requirements.yaml, but missing in charts/ directory: kafka`.
I thought I don't need the kafka dependecy because the condition is `condition: hlf-kafka.enabled`, and in values.yaml it's hlf-kafka.enabled=false.

So I just commented everything out in requirements.yaml.
Now there are no Kafka and Zookeeper instances! :smile: 

Unfortunately, running the first argo channel-flow
`helm template channel-flow/ -f samples/cross-cluster-raft-tls/cluster-one/network.yaml -f samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml -f samples/cross-cluster-raft-tls/cluster-one/hostAliases.yaml | argo submit - --watch`
already doesn't work.



I tried it with both helm versions mentioned above, both k8s cluster were version 1.15.9, and kubectl 1.15.5.

When you tried it out, did you actually clone use the current master branch? Or did you already have some local modifications?
Maybe outcommenting in requirements.yaml already did the job, because all Pods are running, and my issue now is something with argo?

**Comment by raftAtGit** [comment]
thanks for the nice words :)

right, `hlf-kafka.enabled` is in `fabric-kube/hlf-kube/values.yaml`

regardless of the value of `hlf-kafka.enabled` you should make a `helm dependency update`, because  it's a chart dependency. anyway, commenting out `requirements.yaml` should also work.

what error do you get when running channel Argo flow?

**Comment by mr-rooftop** [comment]
Have to admit that I only ever worked with argo when I tried out your repo, so maybe I misconfigured something on this point.

To run argo I executed these commands as suggested by argo's github page:
`kubectl create namespace argo
kubectl apply -n argo -f https://raw.githubusercontent.com/argoproj/argo/stable/manifest/install.yaml`

Then I run in fabric-kube `helm template channel-flow/ -f samples/cross-cluster-raft-tls/cluster-one/network.yaml -f samples/cross-cluster-raft-tls/cluster-one/crypto-config.yaml -f samples/cross-cluster-raft-tls/cluster-one/hostAliases.yaml | argo submit - --watch`

It's output:
![Screenshot_20200310_153651](https://user-images.githubusercontent.com/47558538/76294214-076f7800-62e5-11ea-8626-8dc023f26ba4.png)


When I run the command for cluster two
`helm template channel-flow/ -f samples/cross-cluster-raft-tls/cluster-two/network.yaml -f samples/cross-cluster-raft-tls/cluster-two/crypto-config.yaml -f samples/cross-cluster-raft-tls/cluster-two/hostAliases.yaml | argo submit - --namespace two --watch`
in cluster two I also get the same:
![Screenshot_20200310_163559](https://user-images.githubusercontent.com/47558538/76299211-4275a980-62ed-11ea-82c5-5358069be4d4.png)

**Comment by mr-rooftop** [comment]
Ok, but I just tried out the Scaled-up Raft network with TLS (https://github.com/APGGroeiFabriek/PIVT#scaled-up-raft-network-with-tls) and the part with Argo also failed. 

I **already** got your scaled up Raft network with TLS to run **in the past,** so I'm sure **I'm doing something wrong** with Argo or the helm/k8s versions.


Could you tell me your Argo setup, k8s and helm versions, and HF binary versions so I could mirror them?

I can also open a new issue about my argo problems and close this one.

**Comment by raftAtGit** [comment]
sorry, Argo related issues should not be created this repo. you should check Argo documentation for that.

anyway, you can see why Argo task failed by:
```
argo logs <task name>
```

most possibly you didnt create a service account. check [this page](https://github.com/argoproj/argo/blob/master/docs/getting-started.md#granting-admin-privileges)

**Comment by mr-rooftop** [comment]
42 times thanks to you @raftAtGit , the link solved the problem with argo :pray: 

Thanks also to @LudoNieuw, with both of your help I got the cross-cluster guide to work (using one k8s cluster).

The initial problem that Kafka and Zookeeper Pods got deployed resulted probably in some changes I made myself in the repo, as mentioned by LudoNieuw. **The clean git clone solved this issue!**


For anyone interested, these are my versions and steps I took in addition to the README:
Managed K8s provider: Digital Ocean
K8s server: v1.15.9
Kubectl: v1.15.5
helm: v2.11.0
Hyperledger Fabric binaries: v.1.4.6 installed via `curl -sSL http://bit.ly/2ysbOFE | bash -s -- 1.4.6 1.4.6 0.4.18`
Argo: v2.4.3
jq: 1.5-1
yq: cannot version find right now

To install Tiller I first did `kubectl create -f ./helm-rbac.yaml` with this helm-rbac.yaml:
`apiVersion: v1
kind: ServiceAccount
metadata:
  name: tiller
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: tiller
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - kind: ServiceAccount
    name: tiller
    namespace: kube-system`
and then
`helm init --service-account tiller`

Then I cd into fabric-kube, fabric-kube-two and fabric-kube-three and executed in each directory as said in the README (https://github.com/APGGroeiFabriek/PIVT#launching-the-network):
`helm repo add incubator http://storage.googleapis.com/kubernetes-charts-incubator`
`helm dependency update ./hlf-kube/`

Then I followed the guide.

For the argo commands, I executed the commands as said on their github:
`kubectl create namespace argo`
`kubectl apply -n argo -f https://raw.githubusercontent.com/argoproj/argo/stable/manifests/install.yaml`

And the last magic component was the tip from @raftAtGit to create rolebindings for argo. I created one rolebinding in each namespace:
`kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=default:default`
`kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=two:default  --namespace=two`
`kubectl create rolebinding default-admin --clusterrole=admin --serviceaccount=three:default  --namespace=three`

With that I could successfully run all argo flows until section "Adding new peer organizations" :champagne: 
(one failure in  cluster-two flow invoke-chaincode--even-simpler--private-karga-atlantis--Atlantis--peer1, but maybe it was just a timeout)

Thanks again a lot, you guys are great! :1st_place_medal:
